### PR TITLE
KIBA-1638: Track staking owners

### PR DESCRIPTION
**Pull request by Femi-Ogunkola**
```
curl -X 'POST' \
  'http://localhost:5000/gallery/v1/collections/0xfE8C6d19365453D26af321D0e8c910428c23873F/tokens/query' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "limit": 10,
  "offset": 0,
  "ownerAddress": "0x18090cDA49B21dEAffC21b4F886aed3eB787d032",
  "order": "TOKENID_ASC"
}'
```

### CODE REVIEW

File: api/notd/store/saver.py [hunk 1]
```diff
@@ -1021,3 +1023,39 @@ async def create_gallery_badge_assignment(self, registryAddress: str, ownerAddre
...
+            registryAddress=registryAddress,
+            tokenId=tokenId,
+            stakingDate=stakingDate,
+        )
+
+    async def delete_token_staking(self, tokenStakingId: int, connection: Optional[DatabaseConnection] = None) -> None:
+        query = TokenStakingTable.delete().where(TokenStakingTable.c.tokenStakingId == tokenStakingId).returning(TokenStakingTable.c.tokenStakingId)
+        await self._execute(query=query, connection=connection)
+
+    async def delete_list_token_staking(self, tokenStakingIds: Sequence[int], connection: Optional[DatabaseConnection] = None) -> None:
```
Comment (krishan711) [inline api/notd/store/saver.py:1057]: there shouldnt be list in the name, it should just be plural

File: dbschema/1.tables.sql [hunk 2]
```diff
@@ -482,3 +482,26 @@ CREATE INDEX tbl_gallery_badge_assignments_assigner_address ON tbl_gallery_badge
...
+
+CREATE TABLE tbl_token_stakings (
+    id BIGSERIAL PRIMARY KEY,
+    created_date TIMESTAMP WITHOUT TIME ZONE NOT NULL,
+    updated_date TIMESTAMP WITHOUT TIME ZONE NOT NULL,
+    staking_address TEXT NOT NULL,
+    owner_address TEXT NOT NULL,
+    registry_address TEXT NOT NULL,
+    token_id TEXT NOT NULL,
+    staking_date TIMESTAMP WITHOUT TIME ZONE NOT NULL
```
Comment (krishan711) [inline dbschema/1.tables.sql:494]: staked_date pls (sorry, the tensing is wierd but good to be consistent)

File: dbschema/2.views.sql [hunk 3]
```diff
@@ -39,10 +30,13 @@ CREATE INDEX mvw_user_registry_ordered_ownerships_registry_address_owner_address
 CREATE VIEW vw_token_ownerships AS
 (
     SELECT id, created_date, updated_date, registry_address, token_id, owner_address, transfer_value AS average_transfer_value, transfer_date AS latest_transfer_date, transfer_transaction_hash AS latest_transfer_transaction_hash, 1 AS quantity
-    FROM tbl_token_ownerships
+    FROM tbl_token_ownerships where owner_address != '0xC3503192343EAE4B435E4A1211C5d28BF6f6a696'
```
Comment (krishan711) [inline dbschema/2.views.sql:33]: this isnt great - can we do it in the code instead where needed?

File: api/notd/token_staking_processor.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+from notd.store.schema import TokenTransfersTable
+
+class TokenStakingProcessor:
+
+    def __init__(self, retriever: Retriever) -> None:
+        self.retriever = retriever
+
+    async def calculate_staked_creepz_tokens(self, registryAddress: str) -> List[RetrievedTokenStaking]:
+        stakedQuery = (
+        sqlalchemy.select(TokenTransfersTable.c.tokenId, TokenTransfersTable.c.fromAddress, TokenTransfersTable.c.transactionHash, BlocksTable.c.blockDate)
```
Comment (krishan711) [inline api/notd/token_staking_processor.py:20]: tab in

File: api/notd/staking_manager.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    async def refresh_collection_stakings_deferred(self, address: str) -> None:
+        await self.workQueue.send_message(message=RefreshStakingsForCollectionMessageContent(address=address).to_message())
+
+    async def refresh_collection_stakings(self, address: str) -> None:
+        retrievedTokenStakings = await self.tokenStakingProcessor.calculate_staked_creepz_tokens(registryAddress=address)
+        async with self.saver.create_transaction() as connection:
+            currentTokenStakings = await self.retriever.list_token_stakings(fieldFilters=[
+                StringFieldFilter(fieldName=TokenStakingsTable.c.registryAddress.key, eq=address)
+            ],
+            )
```
Comment (krishan711) [inline api/notd/staking_manager.py:29]:
```suggestion
            ])
```

File: api/notd/token_staking_processor.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+from notd.model import RetrievedTokenStaking
+from notd.model import CREEPZ_STAKING_CONTRACT
+from notd.store.retriever import Retriever
+from notd.store.schema import BlocksTable
+from notd.store.schema import TokenTransfersTable
+
+class TokenStakingProcessor:
+
+    def __init__(self, retriever: Retriever) -> None:
+        self.retriever = retriever
```
Comment (krishan711) [inline api/notd/token_staking_processor.py:16]: i would expect the hardcoded list of staking addresses to be in this class somewhere

File: api/notd/gallery_manager.py [hunk 7]
```diff
@@ -150,13 +153,15 @@ async def query_collection_tokens(self, registryAddress: str, limit: int, offset
...
                 .join(OrderedTokenListingsView, sqlalchemy.and_(TokenMetadatasTable.c.registryAddress == OrderedTokenListingsView.c.registryAddress, TokenMetadatasTable.c.tokenId == OrderedTokenListingsView.c.tokenId, OrderedTokenListingsView.c.tokenListingIndex == 1), isouter=True)
+                .join(TokenStakingsTable, sqlalchemy.and_(TokenOwnershipsView.c.registryAddress == TokenStakingsTable.c.registryAddress, TokenOwnershipsView.c.ownerAddress == TokenStakingsTable.c.ownerAddress, TokenOwnershipsView.c.tokenId == TokenStakingsTable.c.tokenId), isouter=True)
                 .where(TokenMetadatasTable.c.registryAddress == registryAddress)
                 .where(TokenOwnershipsView.c.quantity > 0)
-                .group_by(TokenMetadatasTable.c.tokenMetadataId, TokenCustomizationsTable.c.tokenCustomizationId, OrderedTokenListingsView.c, TokenMetadatasTable.c.registryAddress, TokenMetadatasTable.c.tokenId)
+                .where(TokenOwnershipsView.c.ownerAddress != '0xC3503192343EAE4B435E4A1211C5d28BF6f6a696')
```
Comment (krishan711) [inline api/notd/gallery_manager.py:163]: why do we need to specify this here?

Comment (Femi-Ogunkola) [reply api/notd/gallery_manager.py:163]: Cause If I don't I get multiple tokens on the gallery page, one with the owner address and one with the staking contract?

File: api/notd/api/api_v1.py [hunk 8]
```diff
@@ -116,6 +117,11 @@ async def refresh_latest_listings_for_all_collections_deferred() -> RefreshLates
         await notdManager.refresh_latest_listings_for_all_collections_deferred()
         return RefreshLatestListingsAllCollectionsDeferredResponse()
 
+    @router.post('/collections//refresh-stakings-deferred', response_model=RefreshStakingsDeferredResponse)
```
Comment (krishan711) [inline api/notd/api/api_v1.py:120]:
```suggestion
    @router.post('/collections/refresh-stakings-deferred', response_model=RefreshStakingsDeferredResponse)
```

**Review by krishan711 — CHANGES_REQUESTED** [review]

File: api/notd/model.py [hunk 9]
```diff
@@ -57,6 +57,12 @@
     ],
 }
 
+CREEPZ_STAKING_CONTRACT = '0xC3503192343EAE4B435E4A1211C5d28BF6f6a696'
```
Comment (krishan711) [inline api/notd/model.py:60]: put at the top with others. also should be ADDRESS not CONTRACT

File: api/notd/model.py [hunk 10]
```diff
@@ -57,6 +57,12 @@
...
+
+STAKING_CONTRACTS = [
```
Comment (krishan711) [inline api/notd/model.py:62]: STAKING_ADDRESSES

File: dbschema/2.views.sql [hunk 11]
```diff
@@ -11,27 +11,18 @@ CREATE INDEX mvw_user_registry_first_ownerships_registry_address ON mvw_user_reg
...
-            AND tbl_token_multi_ownerships.registry_address = tbl_token_metadatas.registry_address
-            AND tbl_token_multi_ownerships.quantity > 0
-        WHERE tbl_token_multi_ownerships.registry_address = ANY(array(select registry_address from tbl_gallery_customers))
+        JOIN vw_token_ownerships ON vw_token_ownerships.token_id = tbl_token_metadatas.token_id
+            AND vw_token_ownerships.registry_address = tbl_token_metadatas.registry_address
+            AND vw_token_ownerships.quantity > 0
+        WHERE vw_token_ownerships.registry_address = ANY(array(select registry_address from tbl_gallery_customers))
     )
 );
-CREATE UNIQUE INDEX mvw_user_registry_ordered_ownerships_registry_address_token_id_owner_address ON mvw_user_registry_ordered_ownerships (registry_address, token_id, owner_address);
```
Comment (krishan711) [inline dbschema/2.views.sql:34]: why has this changed?

File: dbschema/2.views.sql [hunk 12]
```diff
@@ -43,6 +34,9 @@ CREATE VIEW vw_token_ownerships AS
 ) UNION (
     SELECT id, created_date, updated_date, registry_address, token_id, owner_address, average_transfer_value, latest_transfer_date, latest_transfer_transaction_hash, quantity
     FROM tbl_token_multi_ownerships
+)UNION (
```
Comment (krishan711) [inline dbschema/2.views.sql:37]: this looks wierd - why do we add this to the ownerships view?

Comment (krishan711) [reply dbschema/2.views.sql:37]: im fairly sure this is bad

File: dbschema/1.tables.sql [hunk 13]
```diff
@@ -482,3 +482,26 @@ CREATE INDEX tbl_gallery_badge_assignments_assigner_address ON tbl_gallery_badge
 CREATE INDEX tbl_gallery_badge_assignments_badge_key ON tbl_gallery_badge_assignments (badge_key);
 CREATE INDEX tbl_gallery_badge_assignments_owner_address ON tbl_gallery_badge_assignments (owner_address);
 CREATE INDEX tbl_gallery_badge_assignments_achieved_date ON tbl_gallery_badge_assignments (achieved_date);
+
+CREATE TABLE tbl_token_stakings (
```
Comment (krishan711) [inline dbschema/1.tables.sql:486]: you need to update the other sql file to grant access to this table

File: api/notd/block_manager.py [hunk 14]
```diff
@@ -66,19 +69,22 @@ async def reprocess_old_blocks(self) -> None:
...
 
-    async def process_block(self, blockNumber: int, shouldSkipProcessingTokens: Optional[bool] = None, shouldSkipUpdatingOwnerships: Optional[bool] = None) -> None:
+    async def process_block(self, blockNumber: int, shouldSkipProcessingTokens: Optional[bool] = None, shouldSkipUpdatingOwnerships: Optional[bool] = None, shouldSkipUpdatingStakings: Optional[bool] = None) -> None:
         processedBlock = await self.blockProcessor.process_block(blockNumber=blockNumber)
         logging.info(f'Found {len(processedBlock.retrievedTokenTransfers)} token transfers in block #{blockNumber}')
         collectionTokenIds = await self._save_processed_block(processedBlock=processedBlock)
         collectionAddresses = list(set(registryAddress for registryAddress, _ in collectionTokenIds))
         logging.info(f'Found {len(collectionTokenIds)} changed tokens and {len(collectionAddresses)} changed collections in block #{blockNumber}')
+        stakingCollectionTokenIds = {(transfer.registryAddress, transfer.tokenId) for transfer in processedBlock.retrievedTokenTransfers if CREEPZ_STAKING_CONTRACT in (transfer.fromAddress, transfer.toAddress)}
```
Comment (krishan711) [inline api/notd/block_manager.py:85]: why creepz hardcoded here? you should refer to the list

File: api/notd/staking_manager.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+            if len(currentTokenStakings) > 0:
+                currentTokenStakingId = currentTokenStakings[0].tokenStakingId
+                logging.info(f'Deleting existing stakings for registryAddress: {registryAddress}, tokenId: {tokenId}')
+                await self.saver.delete_token_staking(tokenStakingId=currentTokenStakingId, connection=connection)
+            if retrievedTokenStaking:
+                logging.info(f'Saving stakings for registryAddress: {registryAddress}, tokenId: {tokenId}')
+                await self.saver.create_token_staking(retrievedTokenStaking=retrievedTokenStaking, connection=connection)
+
+    async def refresh_stakings_for_collections_deferred(self) -> None:
+        await self.workQueue.send_message(message=RefreshStakingsForCollectionMessageContent(address=COLLECTION_CREEPZ_ADDRESS).to_message())
```
Comment (krishan711) [inline api/notd/staking_manager.py:53]: use the list of addresses

File: api/notd/staking_manager.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+
+class StakingManager:
+
+    def __init__(self, retriever: Retriever,saver: Saver, tokenQueue: SqsMessageQueue, workQueue: SqsMessageQueue, tokenStakingProcessor: TokenStakingProcessor) -> None:
+        self.retriever = retriever
+        self.saver = saver
+        self.workQueue = workQueue
+        self.tokenQueue = tokenQueue
+        self.tokenStakingProcessor = tokenStakingProcessor
```
Comment (krishan711) [inline api/notd/staking_manager.py:24]: why is this tokenStakingProcessor but the class here is StakingManager - make both token or both just staking. also the table name, api name, messaage names should all match the same thing

File: api/notd/token_staking_processor.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+
+    async def get_staked_token(self, registryAddress: str, tokenId: str) -> Optional[RetrievedTokenStaking]:
+        for stakingAddress in STAKING_CONTRACTS:
+            latestTokenTransfers = await self.retriever.list_token_transfers(fieldFilters=[
+                    StringFieldFilter(fieldName=TokenStakingsTable.c.registryAddress.key, eq=registryAddress),
+                    StringFieldFilter(fieldName=TokenStakingsTable.c.tokenId.key, eq=tokenId)
+                ],
+            orders=[Order(fieldName=BlocksTable.c.blockDate.key, direction=Direction.DESCENDING)],
+            limit=1)
```
Comment (krishan711) [inline api/notd/token_staking_processor.py:32]:
```suggestion
            latestTokenTransfers = await self.retriever.list_token_transfers(
                fieldFilters=[
                    StringFieldFilter(fieldName=TokenStakingsTable.c.registryAddress.key, eq=registryAddress),
                    StringFieldFilter(fieldName=TokenStakingsTable.c.tokenId.key, eq=tokenId)
                ],
                orders=[Order(fieldName=BlocksTable.c.blockDate.key, direction=Direction.DESCENDING)],
                limit=1,
            )
```

File: api/notd/staking_manager.py (hunk 16 above)
Comment (Femi-Ogunkola) [reply api/notd/staking_manager.py:24]: I just followed the way listings and ownership where done

File: api/application.py [hunk 18]
```diff
@@ -95,12 +97,14 @@
...
 twitterManager = TwitterManager(saver=saver, retriever=retriever, requester=requester, workQueue=workQueue, twitterBearerToken=twitterBearerToken)
 badgeProcessor = BadgeProcessor(retriever=retriever, saver=saver)
 badgeManager = BadgeManager(retriever=retriever, saver=saver, workQueue=workQueue, badgeProcessor=badgeProcessor)
 delegationManager = DelegationManager(ethClient=ethClient)
-notdManager = NotdManager(saver=saver, retriever=retriever, workQueue=workQueue, blockManager=blockManager, tokenManager=tokenManager, activityManager=activityManager, attributeManager=attributeManager, collectionManager=collectionManager, ownershipManager=ownershipManager, listingManager=listingManager, twitterManager=twitterManager, collectionOverlapManager=collectionOverlapManager, badgeManager=badgeManager, delegationManager=delegationManager, requester=requester, revueApiKey=revueApiKey)
+tokenStakingProcessor = TokenStakingProcessor(retriever=retriever)
+stakingManager = StakingManager(retriever=retriever, saver=saver, tokenQueue=tokenQueue, workQueue=workQueue, tokenStakingProcessor=tokenStakingProcessor)
```
Comment (krishan711) [inline api/application.py:105]: name needs updating i think?

File: dbschema/2.views.sql (hunk 11 above)
Comment (krishan711) [reply dbschema/2.views.sql:34]: ping

File: api/notd/block_manager.py [hunk 19]
```diff
@@ -66,19 +69,22 @@ async def reprocess_old_blocks(self) -> None:
         logging.info(f'Scheduling messages for reprocessing {len(blockNumbers)} blocks')
         await self.process_blocks_deferred(blockNumbers=blockNumbers, shouldSkipProcessingTokens=True)
 
-    async def process_blocks_deferred(self, blockNumbers: Sequence[int], shouldSkipProcessingTokens: Optional[bool] = None, delaySeconds: int = 0) -> None:
+    async def  process_blocks_deferred(self, blockNumbers: Sequence[int], shouldSkipProcessingTokens: Optional[bool] = None, delaySeconds: int = 0) -> None:
```
Comment (krishan711) [inline api/notd/block_manager.py:72]: undo

File: api/notd/gallery_manager.py [hunk 20]
```diff
@@ -345,7 +352,6 @@ async def query_collection_users(self, registryAddress: str, order: Optional[str
             sqlalchemy.select(TokenMetadatasTable, UserRegistryOrderedOwnershipsMaterializedView.c.ownerAddress)
                 .join(UserRegistryOrderedOwnershipsMaterializedView, sqlalchemy.and_(UserRegistryOrderedOwnershipsMaterializedView.c.registryAddress == TokenMetadatasTable.c.registryAddress, UserRegistryOrderedOwnershipsMaterializedView.c.tokenId == TokenMetadatasTable.c.tokenId))
                 .where(UserRegistryOrderedOwnershipsMaterializedView.c.registryAddress == registryAddress)
-                .where(UserRegistryOrderedOwnershipsMaterializedView.c.ownerTokenIndex <= 5)
```
Comment (krishan711) [inline api/notd/gallery_manager.py:348]: why remove this?

File: dbschema/2.views.sql [hunk 21]
```diff
@@ -19,22 +21,35 @@ CREATE MATERIALIZED VIEW mvw_user_registry_ordered_ownerships AS (
...
+            WHERE tbl_token_stakings.registry_address = ANY(array(select registry_address from tbl_gallery_customers))
+        )
+    ) as user_registry_ordered_ownerships
+    GROUP BY registry_address, token_id, owner_address
 );
 CREATE UNIQUE INDEX mvw_user_registry_ordered_ownerships_registry_address_token_id_owner_address ON mvw_user_registry_ordered_ownerships (registry_address, token_id, owner_address);
 CREATE INDEX mvw_user_registry_ordered_ownerships_registry_address ON mvw_user_registry_ordered_ownerships (registry_address);
 CREATE INDEX mvw_user_registry_ordered_ownerships_registry_address_token_id ON mvw_user_registry_ordered_ownerships (registry_address, token_id);
-CREATE INDEX mvw_user_registry_ordered_ownerships_registry_address_owner_address ON mvw_user_registry_ordered_ownerships (registry_address, owner_address);
+CREATE INDEX mvw_user_registry_ordered_ownerships_registry_address_owner_address ON mvw_user_registry_ordered_ownerships (registry_address, owner_address);
```
Comment (krishan711) [inline dbschema/2.views.sql:52]: rm empty space

**Review by krishan711 — APPROVED** [review]

**Pull request merged**